src/main/java/io/asyncer/r2dbc/mysql/client/CompressionDuplexCodec.java

```diff
@@ -0,0 +1,243 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.asyncer.r2dbc.mysql.client;
+
+import io.asyncer.r2dbc.mysql.constant.Packets;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandler;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import org.jetbrains.annotations.Nullable;
+
+import java.net.SocketAddress;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A codec that compresses and decompresses packets.
+ * <ul>
+ * <li>Read: compression {@link ByteBuf} -&gt; compression-framed {@link ByteBuf} -&gt;
+ * decompressed {@link ByteBuf}</li>
+ * <li>Write: uncompressed-framed {@link ByteBuf} -&gt; compression-framed {@link ByteBuf}</li>
+ * </ul>
+ */
+final class CompressionDuplexCodec extends ByteToMessageDecoder implements ChannelOutboundHandler {
+
+    static final String NAME = "R2dbcMysqlCompressionDuplexCodec";
+
+    private static final InternalLogger logger =
+        InternalLoggerFactory.getInstance(CompressionDuplexCodec.class);
+
+    private static final int MIN_COMPRESS_LENGTH = 50;
+
+    /**
+     * Compression packet sequence id, incremented independently of the normal sequence id.
+     */
+    private final AtomicInteger sequenceId = new AtomicInteger(0);
+
+    private final Compressor compressor;
+
+    @Nullable
+    private ByteBuf writeCumulated;
+
+    private final Cumulator writeCumulator = MERGE_CUMULATOR;
+
+    private int frameLength = -1;
+
+    CompressionDuplexCodec(Compressor compressor) {
+        this.compressor = compressor;
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        if (msg instanceof ByteBuf) {
+            ByteBuf cumulated = this.writeCumulated == null ? ctx.alloc().buffer(0, 0) :
+                this.writeCumulated;
+
+            this.writeCumulated = cumulated = writeCumulator.cumulate(ctx.alloc(), cumulated, (ByteBuf) msg);
+
+            while (cumulated.readableBytes() >= Packets.MAX_PAYLOAD_SIZE) {
+                logger.trace("Accumulated to the maximum payload, compressing");
+
+                ByteBuf slice = cumulated.readSlice(Packets.MAX_PAYLOAD_SIZE);
+                ByteBuf compressed = compressor.compress(slice);
+
+                if (compressed.readableBytes() >= slice.readableBytes()) {
+                    logger.trace("Sending uncompressed due to compressed payload is larger than original");
+                    compressed.release();
```

Comment:
Compression packets contain two schemas of packets: compressed and uncompressed.

Uncompressed packet is not normal packet, it will contain compression header, normal header, and normal payload. Actually, the entire normal packet is a compression payload of uncompressed packet.

For example:

```pcap
0d 00 00 00 00 00 00 09    00 00 00 03 53 45 4c 45    ............SELE
43 54 20 32                                           CT 2
```

See also: https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_compression_packet.html#sect_protocol_basic_compression_packet_uncompressed_payload

Reply:
Thank you for clarifying that. 

I'll need to check since I've received an error message. It seems like a mistake in my configuration.